pkg/projects/resource_project_user.go

```diff
@@ -0,0 +1,218 @@
+package projects
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/go-resty/resty/v2"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+type User struct {
+	Name  string   `json:"name"`
+	Roles []string `json:"roles"`
+}
+
+type Equatable interface {
+	Equals(other Equatable) bool
+}
+
+func (a User) Equals(b User) bool {
+	return a.Name == b.Name
+}
+
+func contains(as []User, b User) bool {
+	log.Printf("[DEBUG] contains")
+	log.Printf("[TRACE] as: %+v\n", as)
+	log.Printf("[TRACE] b: %+v\n", b)
+
+	for _, a := range as {
+		log.Printf("[TRACE] a: %+v\n", a)
+		log.Printf("[TRACE] a.Equals(b): %+v\n", a.Equals(b))
+		if a.Equals(b) {
+			return true
+		}
+	}
+	return false
+}
+
+var userApply = func(predicate func(bs []User, a User) bool) func(as []User, bs []User) []User {
+	return func(as []User, bs []User) []User {
+		var results []User
+
+		// Not the most efficient way to determine the slices intersection but this suffices for the small-ish number of items
+		for _, a := range as {
+			if predicate(bs, a) {
+				results = append(results, a)
+			}
+		}
+
+		return results
+	}
+}
+
+var usersIntersection = userApply(func(bs []User, a User) bool {
+	return contains(bs, a)
+})
+
+var usersDifference = userApply(func(bs []User, a User) bool {
+	return !contains(bs, a)
+})
+
+type Users struct {
+	Members []User
+}
+
+const projectUsersUrl = "/access/api/v1/projects/%s/users/"
+
+func getMembers(d *ResourceData) []User {
+	var members []User
+
+	if v, ok := d.GetOkExists("member"); ok {
+		projectUsers := v.(*schema.Set).List()
+		if len(projectUsers) == 0 {
+			return members
+		}
+
+		for _, projectUser := range projectUsers {
+			id := projectUser.(map[string]interface{})
+
+			user := User{
+				Name:  id["name"].(string),
+				Roles: castToStringArr(id["roles"].(*schema.Set).List()),
+			}
+			members = append(members, user)
+		}
+	}
+
+	return members
+}
+
+var unpackUsers = func(data *schema.ResourceData) (string, Users, error) {
+	d := &ResourceData{data}
+	projectKey := d.getString("key", false)
+
+	users := Users{
+		Members: getMembers(d),
+	}
+
+	return projectKey, users, nil
+}
+
+var packUsers = func(d *schema.ResourceData, key string, users *[]User) []error {
+	log.Printf("[DEBUG] packUsers")
+
+	setValue := mkLens(d)
+
+	var projectUsers []interface{}
+
+	for _, user := range *users {
+		log.Printf("[TRACE] %+v\n", user)
+		projectUser := map[string]interface{}{
+			"name":  user.Name,
+			"roles": user.Roles,
+		}
+
+		projectUsers = append(projectUsers, projectUser)
+	}
+
+	log.Printf("[TRACE] %s\n", key)
+	log.Printf("[TRACE] %+v\n", projectUsers)
+
+	errors := setValue(key, projectUsers)
+
+	return errors
+}
+
+var getProjectsUsersUrl = func(projectKey string, id string) string {
+	return fmt.Sprintf(projectUsersUrl, projectKey) + id
```

Comment:
Is this necessary? I mean, seems to me that creating the function is more code than not having it? And, you do a `+` on the end of a `Sprintf` - you could just include it. Also, resty has `SetParam("projectKey",projectKey)` and they can be changed (it also has one with a map. 

So, idiomatically, this doesn't match the `Go` idiom or the `Resty` idiom